File: Editorial/101_Symmetric_Tree/TypeScript/2_Iterative.ts
// TypeScript
function isSymmetric(root: TreeNode | null): boolean {
    let q: Array<TreeNode | null> = [root, root];
    while (q.length !== 0) {
        let t1: TreeNode | null = q.shift();
        let t2: TreeNode | null = q.shift();
        if (t1 === null && t2 === null) {
            continue;
        }
        if (t1 === null || t2 === null) {
            return false;
        }
        if (t1.val !== t2.val) {
            return false;
        }
        q.push(t1.left, t2.right, t1.right, t2.left);
    }
    return true;
}

File: Editorial/101_Symmetric_Tree/C/2_Iterative.c
// C
bool isSymmetric(struct TreeNode* root) {
    struct TreeNode* q[3030];
    int head = 0;
    int tail = 0;
    q[head++] = root;
    q[head++] = root;
    while (head != tail) {
        struct TreeNode* t1 = q[tail++];
        struct TreeNode* t2 = q[tail++];
        if (t1 == NULL && t2 == NULL) continue;
        if (t1 == NULL || t2 == NULL) return false;
        if (t1->val != t2->val) return false;
        q[head++] = t1->left;
        q[head++] = t2->right;
        q[head++] = t1->right;
        q[head++] = t2->left;
    }
    return true;
}
